in ElasticKnnSearch added back create_index, add_texts, from_texts

jeffvestal opened the pull request:
Fixes: 
https://github.com/hwchase17/langchain/issues/7117
https://github.com/hwchase17/langchain/issues/5760

Adding back `create_index` , `add_texts`, `from_texts` to ElasticKnnSearch

`from_texts` matches standard `from_texts` methods as quick start up method

`knn_search`  and `hybrid_result` return a list of [`Document()`, `score`,]

# Test `from_texts` for quick start
```
# create new index using from_text

from langchain.vectorstores.elastic_vector_search import ElasticKnnSearch
from langchain.embeddings import ElasticsearchEmbeddings

model_id = "sentence-transformers__all-distilroberta-v1" 
dims = 768
es_cloud_id = ""
es_user = ""
es_password = ""
test_index = "knn_test_index_305"

embeddings = ElasticsearchEmbeddings.from_credentials(
    model_id,
    #input_field=input_field,
    es_cloud_id=es_cloud_id,
    es_user=es_user,
    es_password=es_password,
)

# add texts and create class instance
texts = ["This is a test document", "This is another test document"]
knnvectorsearch = ElasticKnnSearch.from_texts(
    texts=texts,
    embedding=embeddings,
    index_name= test_index,
    vector_query_field='vector',
    query_field='text',
    model_id=model_id,
    dims=dims,
	es_cloud_id=es_cloud_id, 
	es_user=es_user, 
	es_password=es_password
)

# Test `add_texts` method
texts2 = ["Hello, world!", "Machine learning is fun.", "I love Python."]
knnvectorsearch.add_texts(texts2)

query = "Hello"
knn_result = knnvectorsearch.knn_search(query = query, model_id= model_id, k=2)

hybrid_result = knnvectorsearch.knn_hybrid_search(query = query, model_id= model_id, k=2)

```

The  mapping is as follows:
```
{
  "knn_test_index_012": {
    "mappings": {
      "properties": {
        "text": {
          "type": "text"
        },
        "vector": {
          "type": "dense_vector",
          "dims": 768,
          "index": true,
          "similarity": "dot_product"
        }
      }
    }
  }
}
```

# Check response type
```
>>> hybrid_result
[(Document(page_content='Hello, world!', metadata={}), 0.94232327), (Document(page_content='I love Python.', metadata={}), 0.5321523)]

>>> hybrid_result[0]
(Document(page_content='Hello, world!', metadata={}), 0.94232327)

>>> hybrid_result[0][0]
Document(page_content='Hello, world!', metadata={})

>>> type(hybrid_result[0][0])
<class 'langchain.schema.document.Document'>
```

# Test with existing Index
```
from langchain.vectorstores.elastic_vector_search import ElasticKnnSearch
from langchain.embeddings import ElasticsearchEmbeddings

## Initialize ElasticsearchEmbeddings
model_id = "sentence-transformers__all-distilroberta-v1" 
dims = 768
es_cloud_id = 
es_user = ""
es_password = ""
test_index = "knn_test_index_012"

embeddings = ElasticsearchEmbeddings.from_credentials(
    model_id,
    es_cloud_id=es_cloud_id,
    es_user=es_user,
    es_password=es_password,
)

## Initialize ElasticKnnSearch
knn_search = ElasticKnnSearch(
	es_cloud_id=es_cloud_id, 
	es_user=es_user, 
	es_password=es_password, 
	index_name= test_index, 
	embedding= embeddings
)


## Test adding vectors

### Test `add_texts` method when index created
texts = ["Hello, world!", "Machine learning is fun.", "I love Python."]
knn_search.add_texts(texts)

```

jeffvestal committed:
in ElasticKnnSearch added create_index, add_texts, from_texts

jeffvestal committed:
changing create_index to create_knn_index

jeffvestal committed:
Merge branch 'hwchase17:master' into create_index_bug

benwtrent commented on langchain/vectorstores/elastic_vector_search.py:
```diff
@@ -574,3 +574,118 @@ def knn_hybrid_search(
+
+
+
+    def from_texts(
+             self, texts: List[str],
+             dims: int,
+             model_id: Optional[str] = None
+         ) -> None:
```
This should have parity with the parent class. Meaning it should be a `@classmethod` and be used as a ctor for the class, so it should return a `ElasticKnnSearch` not `None`

jeffvestal replied:
I have a separate issue open to return `Document`. I didn't have time this weekend to get that in.

benwtrent commented on langchain/vectorstores/elastic_vector_search.py:
```diff
@@ -574,3 +574,118 @@ def knn_hybrid_search(
+               model_id: Optional[str] = None,
+               refresh_indices: bool = False,
+               **kwargs: Any
+               ) -> None:
+  #             ) -> List[str]:
```
This should effectively override the parent method. So the signature should be:
```
 def add_texts(
        self,
        texts: Iterable[str],
        metadatas: Optional[List[dict]] = None,
        refresh_indices: bool = True,
        ids: Optional[List[str]] = None,
        **kwargs: Any,
    ) -> List[str]:
```

And return generated uuids from indexing if the user didn't provide IDs to index.

jeffvestal replied:
@benwtrent Thanks for the review. 
I was also discussing with @hwchase17 removing `ElasticKnnSearch` as a subclass of `ElasticVectorSearch` and just making it 
```
class ElasticKnnSearch(VectorStore, ABC):
...
```

Since the kNN code doesn't really pull much from the parent class, it seems to just be adding complexity. This may also simplify when we add text_expansion search support

benwtrent replied:
If this overrides the parent or not, it would be really weird for users of the client to have two different methods, named the same thing, with the same goal, but doing different things.

jeffvestal replied:
Totally agree and this was an oversight of mine originally. 
I'm going to fix the two methods up so they behave in the same fashion.

benwtrent reviewed (changes requested):
Something not part of the PR, but we should do, is default to excluding the `vector` field.

`ElasticVectorSearch#similarity_search_with_score`, doesn't specify that it only cares about `metadata` and `text`. So there is unnecessary work and data transfer returning the vectors and any other fields. 

Similarly, I think `ElasticKnnSearch` should by default return `Document` objects like its parent class when doing KNN search, only returning metadata and text. This way users of the Elastic clients have a nice uniform API. Otherwise, it could be weird to switch from `Document` to `dict` and some returning vectors, others not, etc.

jeffvestal committed:
working on lint issues

jeffvestal committed:
reworked some logit to fix linting errors. moved ElasticKnnSearch to VectorStore subclass rather than subclass of ElasticVectorSearch

jeffvestal committed:
Merge branch 'hwchase17:master' into create_index_bug

jeffvestal committed:
returning list of _ids from add_texts

jeffvestal committed:
working on returning Document for knn searches

jeffvestal committed:
fixing lint errors

jeffvestal committed:
saving working changes

jeffvestal committed:
merge conflicts

jeffvestal committed:
finishing cleaning up bugs and adding docstrings

jeffvestal committed:
Merge branch 'create_index_bug' of https://github.com/jeffvestal/langchain into create_index_bug

jeffvestal commented:
@benwtrent do you have time to do another review? I think I addressed all the issues
I removed ElasticKnnSearch as a subclass but tried to align the methods to be standard. I also return Document type. 

I'm also not sure how I picked up 27 other files that show changing

jeffvestal committed:
fixed lint errors

jeffvestal committed:
removed commented out block

jeffvestal commented:
@baskaryan Somehow I picked up 27 other files to change in this PR. Are you able to take a look? It should just be the `langchain/vectorstores/elastic_vector_search.py` file

jeffvestal committed:
Merge branch 'hwchase17:master' into create_index_bug

jeffvestal commented:
Langchain repo underwent a large reorg. 
Closing this PR in favor of https://github.com/langchain-ai/langchain/pull/8180
